Exercício 05, Lista 03

Problema relaxado - Programação Linear (cvxopt.modeling)

In [2]:
from cvxopt.modeling import op, variable
from plotly import graph_objects as go

max f = 3x + 7y \
Sujeito a: \
x <= 3.5 \
5x - 4y <= 10 \
(4/7)x + 2y <= 9 \
x, y >= 0 \


In [2]:
# Definindo variáveis de decisão:
x = variable(1, 'variável x')
y = variable(1, 'variável y')

In [3]:
# Definindo a FOB
fob = -3 * x[0] + -7 * y[0]

In [4]:
# Definindo as restrições
restricoes = []

restricoes.append(x[0] <= 3.5)
restricoes.append(5 * x[0] - 4 * y[0] <= 10)
restricoes.append(4/7 * x[0] + 2 * y[0] <= 9)
restricoes.append(x[0] >= 0)
restricoes.append(y[0] >= 0)

In [5]:
# Montando o PL
problema = op(fob,restricoes)
problema.solve('dense', 'glpk')

In [6]:
print(f'O valor da fob é {-fob.value()[0]}')
print(f'{x.name} é de {x[0].value()[0]}')
print(f'{y.name} é de {y[0].value()[0]}')

O valor da fob é 35.0
variável x é de 3.5
variável y é de 3.5


Programação Linear Inteira (cvxopt.glpk)

In [7]:
from cvxopt.glpk import ilp
import numpy as np
from cvxopt import matrix

In [9]:
4/7

0.5714285714285714

In [11]:
c = matrix([-3, -7], tc = 'd')
G = matrix([[1, 0], [5, -4], [4/7, 2]], tc = 'd')
h = matrix([3.5, 10, 9], tc = 'd')

In [12]:
(status, x) = ilp(c, G.T, h, I=set([0,1]))

In [13]:
fob = sum(c.T * x)

In [14]:
print(status)

optimal


In [17]:
print(f'O valor da fob é {-fob}')
print(f'O valor de x é {x[0]}')
print(f'O valor de y é {x[1]}')

O valor da fob é 31.0
O valor de x é 1.0
O valor de y é 4.0


Programação Linear Inteira (Pulp)

In [19]:
!pip install pulp

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.3/14.3 MB 73.1 MB/s eta 0:00:00


In [20]:
import pulp

In [21]:
# Criação do problema de maximização
problem = pulp.LpProblem("Problema teste", pulp.LpMaximize)

/usr/local/lib/python3.10/dist-packages/pulp/pulp.py:1352: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


In [22]:
# Definindo variáveis de decisão
x = pulp.LpVariable("x", lowBound=0, cat="Integer")
y = pulp.LpVariable("y", lowBound=0, cat="Integer")

In [23]:
# Função Objetivo
problem += 3 * x + 7 * y, "Lucro Total"

In [24]:
# Restrições
problem += x <= 3.5
problem += 5 * x - 4 * y <= 10
problem += 4/7 * x + 2 * y <= 9

In [25]:
# Resolvendo o problema
problem.solve()

1

In [27]:
print(f'Status da solução: {pulp.LpStatus[problem.status]}')
print(f'O valor da fob é: {pulp.value(problem.objective)}')
print(f'O valor de x é: {x.varValue}')
print(f'O valor de y é: {y.varValue}')

Status da solução: Optimal
O valor da fob é: 31.0
O valor de x é: 1.0
O valor de y é: 4.0



---

Solução óptima do problema de PL relaxado: \
F(3.5, 3.5) = 35
- O método cita que é necessário escolher variável que é mais particionada, mais próxima do meio, 0.5, neste caso como as variáveis são iguais, decidi optar pela variável x, mas poderia ser y. \

---

Subproblema A: \

max f = 3x + 7y \

Sujeito a: \

x <= 3.5 \
5x - 4y <= 10 \
(4/7)x + 2y <= 9 \
x <= 3 \
x, y >= 0

---

Subproblema B: \

max f = 3x + 7y \

Sujeito a: \

x <= 3.5 \
5x - 4y <= 10 \
(4/7)x + 2y <= 9 \
x >= 4 \
x, y >= 0


Tentando resolver o problema de forma relaxado \
Branch and Bound - Subproblema A

In [30]:
# Definindo variáveis de decisão:
x = variable(1, 'variável x')
y = variable(1, 'variável y')

# Definindo a FOB
fob = -3 * x[0] + -7 * y[0]

# Definindo as restrições
restricoes = []

restricoes.append(x[0] <= 3.5)
restricoes.append(5 * x[0] - 4 * y[0] <= 10)
restricoes.append(4/7 * x[0] + 2 * y[0] <= 9)
restricoes.append(x[0] >= 0)
restricoes.append(y[0] >= 0)
restricoes.append(x[0] <= 3) # Subproblema A

# Montando o PL
problema = op(fob,restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
  print('Solução ótima encontrada.')
  print(f'O valor da fob é {-fob.value()[0]}')
  print(f'{x.name} é de {x[0].value()[0]}')
  print(f'{y.name} é de {y[0].value()[0]}')
elif problema.status == 'primal infeasible':
  print('O problema é inviável.')
elif problema.status == 'unbounded':
  print('O problema é ilimitado.')
else:
  print('Status desconhecido da otimização.')

optimal
Solução ótima encontrada.
O valor da fob é 34.5
variável x é de 3.0
variável y é de 3.642857142857143


Branch and Bound - Subproblema B

In [31]:
# Definindo variáveis de decisão:
x = variable(1, 'variável x')
y = variable(1, 'variável y')

# Definindo a FOB
fob = -3 * x[0] + -7 * y[0]

# Definindo as restrições
restricoes = []

restricoes.append(x[0] <= 3.5)
restricoes.append(5 * x[0] - 4 * y[0] <= 10)
restricoes.append(4/7 * x[0] + 2 * y[0] <= 9)
restricoes.append(x[0] >= 0)
restricoes.append(y[0] >= 0)
restricoes.append(x[0] >= 4) # Subproblema B

# Montando o PL
problema = op(fob,restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
  print('Solução ótima encontrada.')
  print(f'O valor da fob é {-fob.value()[0]}')
  print(f'{x.name} é de {x[0].value()[0]}')
  print(f'{y.name} é de {y[0].value()[0]}')
elif problema.status == 'primal infeasible':
  print('O problema é inviável.')
elif problema.status == 'unbounded':
  print('O problema é ilimitado.')
else:
  print('Status desconhecido da otimização.')

primal infeasible
O problema é inviável.


Pegando o subproblema A e desmembrando: \
f = (3, 3.642857142857143) = 34.5

--

Subproblema A1: \

max f = 3x + 7y \

Sujeito a: \

x <= 3.5 \
5x - 4y <= 10 \
(4/7)x + 2y <= 9 \
x <= 3 \
y <= 4 \
x, y >= 0

---

Subproblema A2: \

max f = 3x + 7y \

Sujeito a: \

x <= 3.5 \
5x - 4y <= 10 \
(4/7)x + 2y <= 9 \
x <= 3 \
y >= 3 \
x, y >= 0

Subproblema A1:

In [8]:
# Definindo variáveis de decisão:
x = variable(1, 'variável x')
y = variable(1, 'variável y')

# Definindo a FOB
fob = -3 * x[0] + -7 * y[0]

# Definindo as restrições
restricoes = []

restricoes.append(x[0] <= 3.5)
restricoes.append(5 * x[0] - 4 * y[0] <= 10)
restricoes.append(4/7 * x[0] + 2 * y[0] <= 9)
restricoes.append(x[0] >= 0)
restricoes.append(y[0] >= 0)
restricoes.append(x[0] <= 3) # Subproblema A
restricoes.append(y[0] <= 4) # Subproblema A1

# Montando o PL
problema = op(fob,restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
  print('Solução ótima encontrada.')
  print(f'O valor da fob é {-fob.value()[0]}')
  print(f'{x.name} é de {x[0].value()[0]}')
  print(f'{y.name} é de {y[0].value()[0]}')
elif problema.status == 'primal infeasible':
  print('O problema é inviável.')
elif problema.status == 'unbounded':
  print('O problema é ilimitado.')
else:
  print('Status desconhecido da otimização.')

optimal
Solução ótima encontrada.
O valor da fob é 34.5
variável x é de 3.0000000000000004
variável y é de 3.642857142857143


Subproblema A2:

In [9]:
# Definindo variáveis de decisão:
x = variable(1, 'variável x')
y = variable(1, 'variável y')

# Definindo a FOB
fob = -3 * x[0] + -7 * y[0]

# Definindo as restrições
restricoes = []

restricoes.append(x[0] <= 3.5)
restricoes.append(5 * x[0] - 4 * y[0] <= 10)
restricoes.append(4/7 * x[0] + 2 * y[0] <= 9)
restricoes.append(x[0] >= 0)
restricoes.append(y[0] >= 0)
restricoes.append(x[0] <= 3) # Subproblema A
restricoes.append(y[0] >= 3) # Subproblema A2

# Montando o PL
problema = op(fob,restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
  print('Solução ótima encontrada.')
  print(f'O valor da fob é {-fob.value()[0]}')
  print(f'{x.name} é de {x[0].value()[0]}')
  print(f'{y.name} é de {y[0].value()[0]}')
elif problema.status == 'primal infeasible':
  print('O problema é inviável.')
elif problema.status == 'unbounded':
  print('O problema é ilimitado.')
else:
  print('Status desconhecido da otimização.')

optimal
Solução ótima encontrada.
O valor da fob é 34.5
variável x é de 3.0
variável y é de 3.642857142857143


Pegando o subproblema A2 e desmembrando: \
f = (3, 3.642857142857143) = 34.5

--

Subproblema A3_1: \

max f = 3x + 7y \

Sujeito a: \

x <= 3.5 \
5x - 4y <= 10 \
(4/7)x + 2y <= 9 \
x <= 3 \
y >= 3 \
y <= 3 \
x, y >= 0

---

Subproblema A3_2: \

max f = 3x + 7y \

Sujeito a: \

x <= 3.5 \
5x - 4y <= 10 \
(4/7)x + 2y <= 9 \
x <= 3 \
y >= 3 \
y >= 4 \
x, y >= 0

Subproblema A3_1:

In [3]:
# Definindo variáveis de decisão:
x = variable(1, 'variável x')
y = variable(1, 'variável y')

# Definindo a FOB
fob = -3 * x[0] + -7 * y[0]

# Definindo as restrições
restricoes = []

restricoes.append(x[0] <= 3.5)
restricoes.append(5 * x[0] - 4 * y[0] <= 10)
restricoes.append(4/7 * x[0] + 2 * y[0] <= 9)
restricoes.append(x[0] >= 0)
restricoes.append(y[0] >= 0)
restricoes.append(x[0] <= 3) # Subproblema A
restricoes.append(y[0] >= 3) # Subproblema A2
restricoes.append(y[0] <= 3) # Subproblema A3_1

# Montando o PL
problema = op(fob,restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
  print('Solução ótima encontrada.')
  print(f'O valor da fob é {-fob.value()[0]}')
  print(f'{x.name} é de {x[0].value()[0]}')
  print(f'{y.name} é de {y[0].value()[0]}')
elif problema.status == 'primal infeasible':
  print('O problema é inviável.')
elif problema.status == 'unbounded':
  print('O problema é ilimitado.')
else:
  print('Status desconhecido da otimização.')

optimal
Solução ótima encontrada.
O valor da fob é 30.0
variável x é de 3.0
variável y é de 3.0


Subproblema A3_2:

In [4]:
# Definindo variáveis de decisão:
x = variable(1, 'variável x')
y = variable(1, 'variável y')

# Definindo a FOB
fob = -3 * x[0] + -7 * y[0]

# Definindo as restrições
restricoes = []

restricoes.append(x[0] <= 3.5)
restricoes.append(5 * x[0] - 4 * y[0] <= 10)
restricoes.append(4/7 * x[0] + 2 * y[0] <= 9)
restricoes.append(x[0] >= 0)
restricoes.append(y[0] >= 0)
restricoes.append(x[0] <= 3) # Subproblema A
restricoes.append(y[0] >= 3) # Subproblema A2
restricoes.append(y[0] >= 4) # Subproblema A3_2

# Montando o PL
problema = op(fob,restricoes)
problema.solve('dense', 'glpk')

# Verificando o status da otimização
print(problema.status)
if problema.status == 'optimal':
  print('Solução ótima encontrada.')
  print(f'O valor da fob é {-fob.value()[0]}')
  print(f'{x.name} é de {x[0].value()[0]}')
  print(f'{y.name} é de {y[0].value()[0]}')
elif problema.status == 'primal infeasible':
  print('O problema é inviável.')
elif problema.status == 'unbounded':
  print('O problema é ilimitado.')
else:
  print('Status desconhecido da otimização.')

optimal
Solução ótima encontrada.
O valor da fob é 33.25
variável x é de 1.75
variável y é de 4.0
